### ISSUE TITLE
Adding dependency to current rule in `pre_build` fails.

### ISSUE THREAD

**Issue by cemeceme** [thought-machine/please#3318]
If a dependency is added within a `pre_build` function of the current rule, then please doesn't wait for the new dependency to be built before proceeding. 
However, running please a second time, with the dependency now cached, works fine.

This is a similar issue as #3168, since please misses some blocking functionality within the [buildTarget](https://github.com/thought-machine/please/blob/9ae1c966bd53b88d6fd76b396a120568f2d35a8b/src/build/build_step.go#L160C6-L160C17) function for pre and post build calls for cases where dependencies may change.

Ideally, adding new dependencies to the current rule should be permissible in `pre_build`, since there are cases where the sources are not fully known until build time.
Specifically, this is something I noticed when using please to generate code that is then compiled in a separate rule. In that case, the cc-rules for instance, will only use a single build step to compile a potentially large number of sources, running on a single thread. Instead, it should be possible to go through the sources in a `pre_build` rule, to generate rules according to the actual files input.

**Comment by peterebden** [comment]
I don't think this is intended to work at present. I thought we checked for this but I guess apparently not. Basically it runs the pre-build function at the point it's about to build a target - at that point it isn't expecting to back off and build something else. Conceptually it could be made to do that but I don't think that's extremely easy - e.g. the building target currently is taking up one of the build workers, it really needs to release that if it's going to wait for another thing to build, but I don't think there's a good way of handling that right now.

I see the point with the generated code situation - I've observed that before too, and it would be nice if this were possible.

**Comment by cemeceme** [comment]
I dont know if it's that difficult, at least the way I got it to [work on my fork](https://github.com/cemeceme/please/tree/pre-post-build-dependency-handling).
All I'm really doing, is to just wait for all the dependencies to build after the [pre](https://github.com/cemeceme/please/blob/dc4857e44e21deb64e59514b8bb4992735792b0c/src/build/build_step.go#L190)-(and in this fork, also the [post](https://github.com/cemeceme/please/blob/dc4857e44e21deb64e59514b8bb4992735792b0c/src/build/build_step.go#L357)-)build functions.
If these are not defined, there should be no extra overhead, though its been a while since I wrote the code for [dependency registration](https://github.com/cemeceme/please/blob/dc4857e44e21deb64e59514b8bb4992735792b0c/src/core/build_target.go#L1657), so I don't remember how the program flow was.

So far, this seems to work pretty well, and I can write rules that can expand build label outputs at pre-build time.
